File: dotnet/src/Microsoft.Agents.Validation/Validators/MessageValidators/ToolMessageValidator.cs
using Microsoft.Agents.Abstractions.Models;

namespace Microsoft.Agents.Validation.Validators.MessageValidators;

/// <summary>
/// Validator for ToolMessage.
/// </summary>
public class ToolMessageValidator : MessageValidatorBase<ToolMessage>
{
    protected override List<ValidationError> ValidateSpecificFields(ToolMessage message, ValidationContext? context)
    {
        var errors = new List<ValidationError>();

        // MSG-012: ToolMessage must have call-id and name
        if (string.IsNullOrWhiteSpace(message.CallId))
        {
            errors.Add(CreateError(
                ValidationErrorCode.MSG_012,
                "ToolMessage must have a call-id",
                "CallId"));
        }

        if (string.IsNullOrWhiteSpace(message.Name))
        {
            errors.Add(CreateError(
                ValidationErrorCode.MSG_012,
                "ToolMessage must have a name",
                "Name"));
        }

        // REL-005, REL-007, REL-008: Cross-validate with function calls if context available
        // This should happen regardless of whether there are contents
        if (context != null && !string.IsNullOrWhiteSpace(message.CallId))
        {
            var functionCall = context.GetFunctionCall(message.CallId);
            if (functionCall == null)
            {
                errors.Add(CreateError(
                    ValidationErrorCode.REL_005,
                    $"ToolMessage call-id '{message.CallId}' does not match any FunctionCallContent in preceding AgentMessage",
                    "CallId"));
            }
            else if (!string.IsNullOrWhiteSpace(message.Name) &&
                     !string.IsNullOrWhiteSpace(functionCall.Name) &&
                     message.Name != functionCall.Name)
            {
                errors.Add(CreateError(
                    ValidationErrorCode.REL_008,
                    $"ToolMessage name '{message.Name}' does not match FunctionCallContent name '{functionCall.Name}' for call-id '{message.CallId}'",
                    "Name"));
            }
        }

        // ROLE-004: ToolMessage can only contain FunctionResultContent, ErrorContent
        // ROLE-008: FunctionResultContent can only appear in ToolMessage
        if (message.Contents != null && message.Contents.Count > 0)
        {
            var invalidContent = message.Contents
                .Where(c => c is not FunctionResultContent && c is not ErrorContent)
                .ToList();

            if (invalidContent.Any())
            {
                var contentTypes = string.Join(", ", invalidContent.Select(c => c.GetType().Name));
                errors.Add(CreateError(
                    ValidationErrorCode.ROLE_004,
                    $"ToolMessage can only contain FunctionResultContent or ErrorContent, found: {contentTypes}",
                    "Contents"));
            }

            // Validate content items
            errors.AddRange(ValidateContentItems(message.Contents, context));
        }

        return errors;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/message/ToolMessageValidator.ts
import type { ToolMessage } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { MessageValidatorBase } from '../base/MessageValidatorBase';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for ToolMessage.
 */
export class ToolMessageValidator extends MessageValidatorBase<ToolMessage> {
  protected validateSpecificFields(message: ToolMessage, context?: ValidationContext): ValidationError[] {
    const errors: ValidationError[] = [];

    // MSG-012: Tool messages must have toolCallId
    const callIdError = this.validateNotEmpty(
      message.toolCallId,
      'toolCallId',
      ValidationErrorCode.MSG_012,
      'ToolMessage must have toolCallId'
    );
    if (callIdError) {
      errors.push(callIdError);
    }

    // REL-005: toolCallId must match a FunctionCall in preceding AgentMessage
    if (context && message.toolCallId) {
      if (!context.hasFunctionCall(message.toolCallId)) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.REL_005,
            `ToolMessage toolCallId does not match any FunctionCall: ${message.toolCallId}`,
            'toolCallId'
          )
        );
      }
    }

    // ROLE-008: ToolMessage should have exactly one function-result or error content
    if (message.contents) {
      const resultCount = message.contents.filter(
        (c) => c.kind === 'functionResult' || c.kind === 'error'
      ).length;
      if (resultCount === 0) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.ROLE_008,
            'ToolMessage should have at least one function-result or error content',
            'contents'
          )
        );
      }
    }

    return errors;
  }

  protected getAllowedContentTypes(): string[] {
    // ROLE-004: Tool messages can only contain function-result and error
    return ['functionResult', 'error'];
  }

  protected getRoleSpecificErrorCode(): string {
    return ValidationErrorCode.ROLE_004;
  }

  protected getMessageTypeName(): string {
    return 'ToolMessage';
  }

  protected requiresContent(): boolean {
    return true;
  }
}
